### DISCUSSION TITLE
Enable IsDev in tests.NewTestApp()

### DISCUSSION THREAD

**Discussion by iskaa02** [pocketbase/pocketbase#5952] [Q&A] [answered]
i have written some tests for a hook from using the examples at https://pocketbase.io/docs/go-testing/, some tests fail with a 500 status, nothing was printed on the console, i assumed it was a panic that was recovered by the PanicRecoverMiddleware, i tried to unbind it but it didn't work.
I had to use a debugger (i rarely use it) to find it out what was going on, and indeed it was a nil pointer panic.

what is ask for is either set app.isDev to true for test or give the option to set it

**Comment by ganigeorgiev** [comment]
It is intentionally disabled because it is too noisy since each executed query and log will be printed in the console making it impossible to search for the actual test error (especially if you use `t.Parallel()`).

> i tried to unbind it but it didn't work.

You could try something like:

```go
setupTestApp := func(t testing.TB) *tests.TestApp {
    testApp, err := tests.NewTestApp(testDataDir)
    if err != nil {
        t.Fatal(err)
    }

    testApp.OnServe().BindFunc(func(se *core.ServeEvent) error {
        se.Router.Unbind(apis.DefaultPanicRecoverMiddlewareId)

        return se.Next()
    })

    bindAppHooks(testApp)

    return testApp
}
```

> ... or give the option to set it

It is a little more complicated than just changing a flag because the db logs and other internal functions are loaded/registered only once based on the app config value and if we allow changing it during runtime we'll have to restart/reinit any dependent registration and additionally guard the flag  with a mutex/`atomic.Bool` which may introduce unnecessary overhead. I've added in my todo to eventually consider adding at least a new `TestApp` factory method to allow configuring it during `TestApp` initialization but this for now remains a low priority.

**Comment by iskaa02** [comment]
Thank you, i worked but i have a question.

> ```go
> setupTestApp := func(t testing.TB) *tests.TestApp {
>     testApp, err := tests.NewTestApp(testDataDir)
>     if err != nil {
>         t.Fatal(err)
>     }
> 
>     testApp.OnServe().BindFunc(func(se *core.ServeEvent) error {
>         se.Router.Unbind(apis.DefaultPanicRecoverMiddlewareId)
> 
>         return se.Next()
>     })
> 
>     bindAppHooks(testApp)
> 
>     return testApp
> }
>  ```
what is the difference between unbinding on the `OnServe` method and on the router? because the `Router.Unbind` documentation make it seem like if you want to a route group without a specific middleware

**Reply by ganigeorgiev** [reply, marked as answer]
I don't understand what do you mean with the last part.

`OnServe` is like any other app event hook and it could have one or many handlers attached to it. Calling `OnServe().Unbind(someId)` unbinds/removes one of these handlers. For example (_for simplicity explicit ids are used_):

```go
testApp.OnServe().Bind(&hook.Handler[*core.ServeEvent] {
    Id: "test1",
    Func: func(se *core.ServeEvent)  error {
        print(1)
        return se.Next()
    },
})

testApp.OnServe().Bind(&hook.Handler[*core.ServeEvent] {
    Id: "test2",
    Func: func(se *core.ServeEvent)  error {
        print(2)
        return se.Next()
    },
})

testApp.OnServe().Unbind("test1")

// only "test2" is executed when triggering OnServe
```

The router (including sub groups and the individual routes) mimic the hooks API and `Unbind()` removes the handlers attached to the specific router/subgroup/route. Since the router support nesting (aka. subgroups) the rules for removal are a little more complicated than the regular hooks. For more details you can refer to the godoc:
- https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.1/tools/router#RouterGroup.Unbind
- https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.1/tools/router#Route.Unbind
